### PR TITLE
Added Algo for Dutch flag Algorithm 

### LINKED ISSUE #260: [Feature Request] Dutch flag Algo
i want to add Dutch national flag algorithm nder sortin algorithms 
 it sorts an array of 0's and 1's and 2's
@ajay-dhangar 
please assign me with level 2, gssoct-extd, hacktoberfest

image

### PR DESCRIPTION
**Pull request by shimmer12**
### Description
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue number)

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by shimmer12** [comment]
solved issue #260 
@ajay-dhangar please merge this under suitable labels

**Comment by ajay-dhangar** [comment]
> solved issue #260 @ajay-dhangar please merge this under suitable labels

@shimmer12, please sync your fork with the latest changes from the main repository. I've resolved the `backtracking-questions.md` error at line 56.

Here's what you need to do:
1. Fetch the latest updates from the main repository:
   ```
   git fetch upstream
   git merge upstream/main
   ```
2. Push the merged changes to your fork:
   ```
   git push origin your-branch
   ```
3. After syncing, resolve any remaining issues.

Please don't hesitate to let me know if you encounter any problems.

File: docs/algorithms/sorting-algorithms/Dutch-flag-algo.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    for (int i = 0; i < nums.size(); i++) {
+        cout << nums[i] << " ";
+    }
+    cout << endl;
+    
+    return 0;
+}
+```
+### Summary:
+The Dutch National Flag Algorithm efficiently sorts an array containing only 0's, 1's, and 2's in a single traversal of the array. It achieves optimal time complexity of O(N) and operates in-place, making it an ideal solution for this type of problem. The use of three pointers (low, mid, and high) ensures that the array is correctly partitioned, and the algorithm is both simple and effective.
```
Comment (ajay-dhangar) [inline docs/algorithms/sorting-algorithms/Dutch-flag-algo.md]: replace `O(N)` with `$O(N)$`

File: docs/algorithms/sorting-algorithms/Dutch-flag-algo.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+- **Average Case: O(N)**  
+  The algorithm still runs in linear time even for a randomly arranged array, as each element is processed at most once.
+
+- **Worst Case: O(N)**  
+  In the worst case, the array may require multiple swaps, but the time complexity remains O(N), where N is the size of the array.
+
+### Space Complexity:
+
+- **Space Complexity: O(1)**
```
Comment (ajay-dhangar) [inline docs/algorithms/sorting-algorithms/Dutch-flag-algo.md]: replace `O(1)` with `$O(1)$`

File: docs/algorithms/sorting-algorithms/Dutch-flag-algo.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+### Time Complexity:
+
+- **Best Case: O(N)**  
+  The array is traversed only once using the `mid` pointer, resulting in linear time complexity.
+
+- **Average Case: O(N)**  
+  The algorithm still runs in linear time even for a randomly arranged array, as each element is processed at most once.
+
+- **Worst Case: O(N)**
```
Comment (ajay-dhangar) [inline docs/algorithms/sorting-algorithms/Dutch-flag-algo.md]: replace `O(N)` with `$O(N)$`

File: docs/algorithms/sorting-algorithms/Dutch-flag-algo.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+- **Efficient for Specific Inputs**:  
+  The algorithm is optimized for arrays with only three distinct values (0, 1, and 2), achieving linear time complexity.
+
+### Time Complexity:
+
+- **Best Case: O(N)**  
+  The array is traversed only once using the `mid` pointer, resulting in linear time complexity.
+
+- **Average Case: O(N)**
```
Comment (ajay-dhangar) [inline docs/algorithms/sorting-algorithms/Dutch-flag-algo.md]: replace `O(N)` with `$O(N)$`

File: docs/algorithms/sorting-algorithms/Dutch-flag-algo.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+- **In-Place Sorting**:  
+  The array is sorted in-place without using extra space, making the algorithm space-efficient.
+
+- **Efficient for Specific Inputs**:  
+  The algorithm is optimized for arrays with only three distinct values (0, 1, and 2), achieving linear time complexity.
+
+### Time Complexity:
+
+- **Best Case: O(N)**
```
Comment (ajay-dhangar) [inline docs/algorithms/sorting-algorithms/Dutch-flag-algo.md]: replace `O(N)` with `$O(N)$`

**Review by ajay-dhangar — CHANGES_REQUESTED** [review]
update code based on suggestions

**Comment by shimmer12** [comment]
I have made the required changes
thank you !

**Review by ajay-dhangar — APPROVED** [review]
Done

**Pull request merged**